[NEW_PORT] comms/uhubctl: utility to control usb hub port power.

cederom opened the pull request:
Uhubctl is utility to control USB power per-port on smart USB hubs. Smart hub is defined as one that implements per-port power switching.

Note that quite a few modern motherboards have built-in root hubs that do support this feature - you may not even need to buy any external hub.

This utility was tested to compile and work on Linux (Ubuntu/Debian/Raspbian, Redhat/EPEL/Fedora/CentOS, Arch Linux, Gentoo, openSUSE, Buildroot), FreeBSD, NetBSD, SunOS and MacOS.

WWW: https://github.com/mvp/uhubctl

bsdimp commented:
I have 17 IN 1 Powered USB 3.2 Hub: VANGREE Powered USB Hub and ROSONWAY 16-Port USB Hub,. The first one barely works for status only. Big pain. I should have returned it, but got busy with travel...  The second one mostly works, but  off -> on -> off confuses the USB stack I think.

cederom commented:
This is standard output with two Unitek 7 port USB 3.0 hubs:

```
% uhubctl
Current status for hub 2-0.0 [0000:0000 (0x1033) XHCI root HUB]
  Port 1: 06a0 power Rx.Detect
  Port 2: 0643 power U2 enable connect []
  Port 3: 0403 off enable connect []
  Port 4: 0403 off enable connect []
Current status for hub 2-4 [0bda:5411 Generic 4-Port USB 2.0 Hub]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0503 power highspeed enable connect [0bda:5411 Generic 4-Port USB 2.0 Hub]
Current status for hub 2-4.4 [0bda:5411 Generic 4-Port USB 2.0 Hub]
  Port 1: 0103 power enable connect [044f:b106 Thrustmaster T.Flight Stick X]
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
Current status for hub 2-2 [0bda:0411 Generic 4-Port USB 3.0 Hub]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 0243 power 5gbps U2 enable connect [0bda:0411 Generic 4-Port USB 3.0 Hub]
Current status for hub 2-2.4 [0bda:0411 Generic 4-Port USB 3.0 Hub]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 02a0 power 5gbps Rx.Detect
Current status for hub 1-9 [0bda:5411 Generic 4-Port USB 2.0 Hub]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0503 power highspeed enable connect [046d:082d HD Pro Webcam C920 0D9BB5AF]
  Port 4: 0503 power highspeed enable connect [0bda:5411 Generic 4-Port USB 2.0 Hub]
Current status for hub 1-9.4 [0bda:5411 Generic 4-Port USB 2.0 Hub]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
Current status for hub 1-17 [0bda:0411 Generic 4-Port USB 3.0 Hub]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 0223 power 5gbps U1 enable connect [0bda:0411 Generic 4-Port USB 3.0 Hub]
Current status for hub 1-17.4 [0bda:0411 Generic 4-Port USB 3.0 Hub]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 02a0 power 5gbps Rx.Detect
```

This is with 4xU3CHARGEHUB16 (i-tec USB 3.0 Charging HUB 90W 16x USB 3.0):

```
% uhubctl -f
Current status for hub 2-0.0 [0000:0000 (0x1033) XHCI root HUB]
  Port 1: 06a0 power Rx.Detect
  Port 2: 0643 power U2 enable connect []
  Port 3: 0403 off enable connect []
  Port 4: 0403 off enable connect []
Current status for hub 1-0.0 [0000:0000 Intel XHCI root HUB]
  Port 1: 06a0 power Rx.Detect
  Port 2: 0003 off enable connect []
  Port 3: 0403 off enable connect []
  Port 4: 0003 off enable connect []
  Port 5: 0403 off enable connect []
  Port 6: 06a0 power Rx.Detect
  Port 7: 06a0 power Rx.Detect
  Port 8: 06a0 power Rx.Detect
  Port 9: 0403 off enable connect []
  Port 10: 06a0 power Rx.Detect
  Port 11: 06a0 power Rx.Detect
  Port 12: 06a0 power Rx.Detect
  Port 13: 0403 off enable connect []
  Port 14: 06a0 power Rx.Detect
Current status for hub 0-0.0 [0000:0000 Intel XHCI root HUB]
  Port 1: 06a0 power Rx.Detect
  Port 2: 06a0 power Rx.Detect
  Port 3: 06a0 power Rx.Detect
Current status for hub 2-3 [1a40:0801 USB 2.0 Hub]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0503 power highspeed enable connect [1a40:0801 USB 2.0 Hub]
  Port 4: 0103 power enable connect [1a86:8074 WCH.CN]
Current status for hub 1-4 [0a12:4010]
  Port 1: 0103 power enable connect [0b0e:24ca Jabra Link 380 08C8C22A8936]
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
Current status for hub 2-3.3 [1a40:0801 USB 2.0 Hub]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0503 power highspeed enable connect [1a40:0101 USB 2.0 Hub [MTT]]
Current status for hub 1-5 [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 1: 0503 power highspeed enable connect [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 2: 0503 power highspeed enable connect [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 3: 0503 power highspeed enable connect [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 4: 0503 power highspeed enable connect [05e3:0610 GenesysLogic USB2.1 Hub]
Current status for hub 2-3.3.4 [1a40:0101 USB 2.0 Hub [MTT]]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0303 power lowspeed enable connect [09da:6406 A4tech A4tech KB]
Current status for hub 1-5.1 [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 1: 0503 power highspeed enable connect [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 2: 0100 power
  Port 3: 0103 power enable connect [0483:374b STMicroelectronics STM32 STLink 0670FF323535474B43101734]
  Port 4: 0103 power enable connect [10c4:ea60 Silicon Labs CP2102 USB to UART Bridge Controller 0001]
Current status for hub 1-5.2 [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 1: 0103 power enable connect [10c4:ea60 Silicon Labs ESP32DEVKITC-CeDeROM-76f24aad88225fedf7260bb4f54c525d ESP32DEVKITC-CeDeROM-76f24aad88225fedf7260bb4f54c525d]
  Port 2: 0103 power enable connect [10c4:ea60 Silicon Labs CP2102N USB to UART Bridge Controller 1a2626680675eb11955981afb7be2ba5]
  Port 3: 0101 power connect []
  Port 4: 0100 power
Current status for hub 2-3.4 [1a86:8074 WCH.CN]
  Port 1: 0100 power
  Port 2: 0103 power enable connect [046d:c099 Logitech G502 X 205236675046]
  Port 3: 0100 power
  Port 4: 0100 power
Current status for hub 1-5.3 [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 1: 0101 power connect []
  Port 2: 0101 power connect []
  Port 3: 0101 power connect []
  Port 4: 0101 power connect []
Current status for hub 1-5.4 [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 1: 0100 power
  Port 2: 0103 power enable connect [10c4:ea60 Silicon Labs CP2102N USB to UART Bridge Controller 26ea4a5fd936e81198aa8fbb0f349d82]
  Port 3: 0503 power highspeed enable connect [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 4: 0503 power highspeed enable connect [05e3:0610 GenesysLogic USB2.1 Hub]
Current status for hub 2-4 [0bda:5411 Generic 4-Port USB 2.0 Hub]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0503 power highspeed enable connect [0bda:5411 Generic 4-Port USB 2.0 Hub]
Current status for hub 2-4.4 [0bda:5411 Generic 4-Port USB 2.0 Hub]
  Port 1: 0103 power enable connect [044f:b106 Thrustmaster T.Flight Stick X]
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
Current status for hub 1-13 [05e3:0608 USB2.0 Hub]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
Current status for hub 1-21 [05e3:0626 GenesysLogic USB3.1 Hub]
  Port 1: 0243 power 5gbps U2 enable connect [05e3:0626 GenesysLogic USB3.1 Hub]
  Port 2: 0243 power 5gbps U2 enable connect [05e3:0626 GenesysLogic USB3.1 Hub]
  Port 3: 0243 power 5gbps U2 enable connect [05e3:0626 GenesysLogic USB3.1 Hub]
  Port 4: 0243 power 5gbps U2 enable connect [05e3:0626 GenesysLogic USB3.1 Hub]
Current status for hub 1-21.1 [05e3:0626 GenesysLogic USB3.1 Hub]
  Port 1: 0243 power 5gbps U2 enable connect [05e3:0626 GenesysLogic USB3.1 Hub]
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 02a0 power 5gbps Rx.Detect
Current status for hub 1-21.2 [05e3:0626 GenesysLogic USB3.1 Hub]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 02a0 power 5gbps Rx.Detect
Current status for hub 1-21.3 [05e3:0626 GenesysLogic USB3.1 Hub]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 02a0 power 5gbps Rx.Detect
Current status for hub 1-21.4 [05e3:0626 GenesysLogic USB3.1 Hub]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 0243 power 5gbps U2 enable connect [05e3:0626 GenesysLogic USB3.1 Hub]
Current status for hub 2-2 [0bda:0411 Generic 4-Port USB 3.0 Hub]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 0243 power 5gbps U2 enable connect [0bda:0411 Generic 4-Port USB 3.0 Hub]
Current status for hub 2-2.4 [0bda:0411 Generic 4-Port USB 3.0 Hub]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 02a0 power 5gbps Rx.Detect
Current status for hub 1-9 [0bda:5411 Generic 4-Port USB 2.0 Hub]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0503 power highspeed enable connect [046d:082d HD Pro Webcam C920 0D9BB5AF]
  Port 4: 0503 power highspeed enable connect [0bda:5411 Generic 4-Port USB 2.0 Hub]
Current status for hub 1-9.4 [0bda:5411 Generic 4-Port USB 2.0 Hub]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
Current status for hub 1-17 [0bda:0411 Generic 4-Port USB 3.0 Hub]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 0243 power 5gbps U2 enable connect [0bda:0411 Generic 4-Port USB 3.0 Hub]
Current status for hub 1-17.4 [0bda:0411 Generic 4-Port USB 3.0 Hub]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 02a0 power 5gbps Rx.Detect
Current status for hub 1-21.1.1 [05e3:0626 GenesysLogic USB3.1 Hub]
  Port 1: 0243 power 5gbps U2 enable connect [05e3:0626 GenesysLogic USB3.1 Hub]
  Port 2: 0243 power 5gbps U2 enable connect [05e3:0626 GenesysLogic USB3.1 Hub]
  Port 3: 0243 power 5gbps U2 enable connect [05e3:0626 GenesysLogic USB3.1 Hub]
  Port 4: 0243 power 5gbps U2 enable connect [05e3:0626 GenesysLogic USB3.1 Hub]
Current status for hub 1-21.1.1.1 [05e3:0626 GenesysLogic USB3.1 Hub]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 02a0 power 5gbps Rx.Detect
Current status for hub 1-21.1.1.2 [05e3:0626 GenesysLogic USB3.1 Hub]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 02a0 power 5gbps Rx.Detect
Current status for hub 1-21.1.1.3 [05e3:0626 GenesysLogic USB3.1 Hub]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 02a0 power 5gbps Rx.Detect
Current status for hub 1-21.1.1.4 [05e3:0626 GenesysLogic USB3.1 Hub]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 02a0 power 5gbps Rx.Detect
Current status for hub 1-5.1.1 [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 1: 0503 power highspeed enable connect [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 2: 0503 power highspeed enable connect [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 3: 0503 power highspeed enable connect [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 4: 0503 power highspeed enable connect [05e3:0610 GenesysLogic USB2.1 Hub]
Current status for hub 1-5.1.1.1 [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0103 power enable connect [0483:374b STMicroelectronics STM32 STLink 0672FF373857343143073259]
  Port 4: 0103 power enable connect [0483:374b STMicroelectronics STM32 STLink 0668FF505771664867040932]
Current status for hub 1-5.1.1.2 [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0103 power enable connect [303a:1001 Espressif USB JTAG/serial debug unit 54:32:04:29:3E:00]
Current status for hub 1-5.1.1.3 [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0103 power enable connect [2e8a:8042 Raspberry Pi PicoArduino 4250305031373511]
  Port 4: 0103 power enable connect [0483:374b STMicroelectronics STM32 STLink 066AFF514951775087041740]
Current status for hub 1-5.1.1.4 [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 1: 0103 power enable connect [0d28:0204 MBED MBED CMSIS-DAP 10101f2e2a69434f48b8d3318c55f2091018]
  Port 2: 0103 power enable connect [0483:374b STMicroelectronics STM32 STLink 0672FF535752877167094820]
  Port 3: 0103 power enable connect [0483:374b STMicroelectronics STM32 STLink 066CFF495252717267232538]
  Port 4: 0503 power highspeed enable connect [0483:374e STMicroelectronics STLINK-V3 0012000E544B501320343637]
Current status for hub 1-5.4.3 [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 1: 0503 power highspeed enable connect [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 2: 0503 power highspeed enable connect [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 3: 0503 power highspeed enable connect [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 4: 0503 power highspeed enable connect [05e3:0610 GenesysLogic USB2.1 Hub]
Current status for hub 1-5.4.3.1 [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 1: 0100 power
  Port 2: 0503 power highspeed enable connect [0483:3754 STMicroelectronics STLINK-V3 0026002A3234510233353533]
  Port 3: 0503 power highspeed enable connect [0483:374e STMicroelectronics STLINK-V3 003000223233510D39363634]
  Port 4: 0103 power enable connect [1366:0105 SEGGER J-Link 000591063449]
Current status for hub 1-5.4.3.2 [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 1: 0103 power enable connect [0403:6001 FTDI FT232R USB UART A5069RR4]
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
Current status for hub 1-5.4.3.3 [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
Current status for hub 1-5.4.3.4 [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
Current status for hub 1-5.4.4 [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 1: 0503 power highspeed enable connect [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 2: 0503 power highspeed enable connect [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 3: 0503 power highspeed enable connect [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 4: 0503 power highspeed enable connect [05e3:0610 GenesysLogic USB2.1 Hub]
Current status for hub 1-5.4.4.1 [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 1: 0100 power
  Port 2: 0103 power enable connect [0525:a4a7 NuttX CDC/ACM Serial 0]
  Port 3: 0100 power
  Port 4: 0100 power
Current status for hub 1-5.4.4.2 [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 1: 0103 power enable connect [0525:a4a7 NuttX CDC/ACM Serial 0]
  Port 2: 0103 power enable connect [2886:802f Seeed Seeed femto M0 FD76A4045030524E572E3120FF133110]
  Port 3: 0103 power enable connect [303a:1001 Espressif USB JTAG/serial debug unit E4:B0:63:41:88:A4]
  Port 4: 0103 power enable connect [0d28:0204 ARM DAPLink CMSIS-DAP 4600000017b6370300000000000000000000000097969902]
Current status for hub 1-5.4.4.3 [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
Current status for hub 1-5.4.4.4 [05e3:0610 GenesysLogic USB2.1 Hub]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
Current status for hub 1-21.4.4 [05e3:0626 GenesysLogic USB3.1 Hub]
  Port 1: 0243 power 5gbps U2 enable connect []
  Port 2: 0243 power 5gbps U2 enable connect []
  Port 3: 0243 power 5gbps U2 enable connect []
  Port 4: 0243 power 5gbps U2 enable connect []
```

It seems this itec is not yet supported:

```
% uhubctl -l 1-5.4.4.2 -p 1 -a off
No compatible devices detected at location 1-5.4.4.2!
```

With Unitek it seems to work, except after power off device is not detached, it is detached and re-attached later at power on stage, the same with as with `-a cycle`:

```
% uhubctl -l 2-4.4 -p 1 -a on
Current status for hub 2-4.4 [0bda:5411 Generic 4-Port USB 2.0 Hub]
  Port 1: 0000 off
Sent power on request
New status for hub 2-4.4 [0bda:5411 Generic 4-Port USB 2.0 Hub]
Current status for hub 2-2.4 [0bda:0411 Generic 4-Port USB 3.0 Hub]
  Port 1: 00a0 off
Sent power on request
New status for hub 2-2.4 [0bda:0411 Generic 4-Port USB 3.0 Hub]

% tail -f /var/log/messages
Oct  3 21:48:58 hexagon kernel: ugen2.12: <Thrustmaster T.Flight Stick X> at usbus2 (disconnected)
Oct  3 21:48:58 hexagon kernel: usbhid5: at uhub17, port 1, addr 11 (disconnected)
Oct  3 21:48:58 hexagon kernel: hidraw5: detached
Oct  3 21:48:58 hexagon kernel: hgame0: detached
Oct  3 21:48:58 hexagon kernel: hidbus5: detached
Oct  3 21:48:58 hexagon kernel: usbhid5: detached
Oct  3 21:48:59 hexagon kernel: ugen2.12: <Thrustmaster T.Flight Stick X> at usbus2
Oct  3 21:48:59 hexagon kernel: usbhid5 on uhub17
Oct  3 21:48:59 hexagon kernel: usbhid5: <Thrustmaster T.Flight Stick X, class 0/0, rev 1.10/1.00, addr 11> on usbus2
Oct  3 21:48:59 hexagon kernel: hidbus5: <HID bus> on usbhid5
Oct  3 21:48:59 hexagon kernel: hgame0: <Thrustmaster T.Flight Stick X Joystick> on hidbus5
Oct  3 21:48:59 hexagon kernel: hidraw5: <Thrustmaster T.Flight Stick X Raw HID Device> on hidbus5
```

For me this quick list and power cycle of hanging boards is enough for now. 

On FreeBSD is is always possible to use `usbconfig` and this also works on the itec hub attached devices:

```
# usbconfig
(..)
ugen1.28: <J-Link SEGGER> at usbus1, cfg=0 md=HOST spd=FULL (12Mbps) pwr=ON (100mA)
(..)
ugen2.12: <T.Flight Stick X ThrustMaster, Inc.> at usbus2, cfg=0 md=HOST spd=FULL (12Mbps) pwr=ON (80mA)

# usbconfig -d 2.12 power_off

# tail -f /var/log/messages
Oct  3 21:55:17 hexagon kernel: usbhid5: at uhub17, port 1, addr 11 (disconnected)
Oct  3 21:55:17 hexagon kernel: hidraw5: detached
Oct  3 21:55:17 hexagon kernel: hgame0: detached
Oct  3 21:55:17 hexagon kernel: hidbus5: detached
Oct  3 21:55:17 hexagon kernel: usbhid5: detached

# usbconfig -d 2.12 power_on

# tail -f  /var/log/messages
Oct  3 21:55:27 hexagon kernel: usbhid5 on uhub17
Oct  3 21:55:27 hexagon kernel: usbhid5: <Thrustmaster T.Flight Stick X, class 0/0, rev 1.10/1.00, addr 11> on usbus2
Oct  3 21:55:27 hexagon kernel: hidbus5: <HID bus> on usbhid5
Oct  3 21:55:27 hexagon kernel: hgame0: <Thrustmaster T.Flight Stick X Joystick> on hidbus5
Oct  3 21:55:27 hexagon kernel: hidraw5: <Thrustmaster T.Flight Stick X Raw HID Device> on hidbus5

# usbconfig -d 1.28 power_off

# tail -f  /var/log/messages
Oct  3 21:58:38 hexagon kernel: umodem3: at uhub28, port 4, addr 34 (disconnected)
Oct  3 21:58:38 hexagon kernel: umodem3: detached


# usbconfig -d 1.28 power_on
# tail -f  /var/log/messages
Oct  3 21:58:43 hexagon kernel: umodem0: <SEGGER J-Link, class 239/2, rev 2.00/1.00, addr 34> on usbus1
Oct  3 21:58:43 hexagon kernel: umodem0: data interface 1, has no CM over data, has break
```

This `uhubctl` utility makes usb port control portable and this is most important. I have also dedicated rPI-Zero-2W board for automation testing running FreeBSD, its slower but consumes at most 5W of power :-) But there are other users that probably will prefer Linux thus I need this tool already existing as package even if it need fixes in the upstream :-)

cederom commented:
> I have 17 IN 1 Powered USB 3.2 Hub: VANGREE Powered USB Hub and ROSONWAY 16-Port USB Hub,. The first one barely works for status only. Big pain. I should have returned it, but got busy with travel... The second one mostly works, but off -> on -> off confuses the USB stack I think.

Have you tried `usbconfig power_on/off` @bsdimp ? :-)

arrowd commented on comms/uhubctl/Makefile:
```diff
@@ -0,0 +1,22 @@
+USES=		gmake
+USE_GITHUB=	yes
+GH_ACCOUNT=	mvp
+GH_TAGNAME=	v2.6.0
```
Remove this and instead use `DISTVERSIONPREFIX=v`. There are plenty of ports that can be used as example.

cederom commented on comms/uhubctl/Makefile:
```diff
@@ -0,0 +1,22 @@
+WWW=		https://github.com/mvp/uhubctl
+
+LICENSE=	GPLv2
+LICENSE_FILE	=${WRKSRC}/LICENSE
```
done :-)

cederom commented on comms/uhubctl/pkg-plist:
```diff
@@ -0,0 +1 @@
+sbin/uhubctl
```
done :-)

arrowd commented:
Building on 16-CURRENT yields:

```
cc  -O2 -pipe  -fstack-protector-strong -fno-strict-aliasing  -Wall -Wextra -std=c99 -pedantic -DPROGRAM_VERSION=\"2.6.0\" uhubctl.c -o uhubctl   -lusb-1.0
ld: error: unable to find library -lusb-1.0
cc: error: linker command failed with exit code 1 (use -v to see invocation)
```

cederom commented:
Thanks @arrowd :-)
* Moved to `sysutils/uhubctl` no problem, put initially in the (usb) comms :-P
* Added `USES=pkgconfig` [1] can you please check if that fixes build on 16?
  ```
  % pkgconf --libs libusb-1.0
  -lusb
  ```

[1] https://docs.freebsd.org/en/books/porters-handbook/uses/index.html#uses-pkgconfig

arrowd commented:
Merged, thanks.

cederom commented:
TANK U SIR! =)
